### PR TITLE
Stack borrows: weak protectors

### PR DESCRIPTION
**Pull request by RalfJung**
This addresses the issue described in https://github.com/rust-lang/unsafe-code-guidelines/issues/376.

### CODE REVIEW

**Comment by RalfJung** [comment]
See https://github.com/rust-lang/unsafe-code-guidelines/pull/377 for a conceptual description of this change.

**Comment by RalfJung** [comment]
At least the liballoc tests all this pass with this patch. I doubt there will be a lot of breakage out there, but I guess @saethlin will see. ;)

File: src/stacked_borrows/mod.rs [hunk 1]
```diff
@@ -377,7 +401,7 @@ impl<'tcx> Stack {
                 0
             };
             self.pop_items_after(first_incompatible_idx, |item| {
-                Stack::item_popped(&item, global, dcx)?;
+                Stack::item_popped(&item, global, dcx, /* deallocation */ false)?;
```
Comment (saethlin) [inline src/stacked_borrows/mod.rs:404]: I think when you feel the need to put a comment next to a bool like this, a bool is no longer the best choice. Could this be two functions now? Or an enum, perhaps:
```rust
enum PopCause {
    Deallocation,
    Aliasing,
}
```
(I am not excited by these names)

File: src/stacked_borrows/diagnostics.rs [hunk 2]
```diff
@@ -288,7 +288,11 @@ impl<'span, 'history, 'ecx, 'mir, 'tcx> DiagnosticCx<'span, 'history, 'ecx, 'mir
             }
             Operation::Access(AccessOp { kind, range, .. }) =>
                 (*range, InvalidationCause::Access(*kind)),
-            _ => unreachable!("Tags can only be invalidated during a retag or access"),
+            _ => {
+                // This can be reached, but never be relevant later since the entire allocation is
+                // gone now.
+                return;
```
Comment (saethlin) [inline src/stacked_borrows/diagnostics.rs:294]: Without the runtime assertion, I would be happier with the variant we are ignoring listed explicitly so this is less fragile under refactoring.

File: src/stacked_borrows/mod.rs (hunk 1 above)
Comment (RalfJung) [reply src/stacked_borrows/mod.rs:404]: Yeah, we have some of these and I'm just usually too lazy... I wish we had named function parameters.^^

**Comment by saethlin** [comment]
I'm happy with the implementation. Thanks for cleaning up all the unnecessary mutability along the way.

**Pull request merged**